File: Java_Angular_Typescript/02 - Challenge/Resolved/back/src/main/java/deep/session/CA_CIB/service/MovieService.java
package deep.session.CA_CIB.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import deep.session.CA_CIB.exception.MovieAlreadyExistsException;
import deep.session.CA_CIB.exception.MovieNotFoundException;
import deep.session.CA_CIB.model.Movie;
import deep.session.CA_CIB.repository.MovieRepository;

import java.util.List;

@Service
public class MovieService {
    @Autowired
    private MovieRepository movieRepository;

    public List<Movie> getAllMovies() {
        return movieRepository.findAll();
    }

    public Movie getMovieById(Long id) {
        return movieRepository.findById(id).orElseThrow(() -> new MovieNotFoundException("Movie not found with id: " + id));
    }

    @Transactional
    public Movie saveMovie(Movie movie) {
        if (movieRepository.existsByTitleAndDirector(movie.getTitle(), movie.getDirector())) {
            throw new MovieAlreadyExistsException("Movie already exists with title: " + movie.getTitle() + " and director: " + movie.getDirector());
        }
        return movieRepository.save(movie);
    }

    @Transactional
    public Movie updateMovie(Long id, Movie movie) {
        Movie existingMovie = movieRepository.findById(id).orElseThrow(() -> new MovieNotFoundException("Movie not found with id: " + id));
        if (!existingMovie.getTitle().equals(movie.getTitle()) || !existingMovie.getDirector().equals(movie.getDirector())) {
            if (movieRepository.existsByTitleAndDirector(movie.getTitle(), movie.getDirector())) {
                throw new MovieAlreadyExistsException("Movie already exists with title: " + movie.getTitle() + " and director: " + movie.getDirector());
            }
        }
        existingMovie.setTitle(movie.getTitle());
        existingMovie.setDirector(movie.getDirector());
        existingMovie.setReleaseYear(movie.getReleaseYear());
        return movieRepository.save(existingMovie);
    }

    @Transactional
    public void deleteMovie(Long id) {
        if (!movieRepository.existsById(id)) {
            throw new MovieNotFoundException("Movie not found with id: " + id);
        }
        movieRepository.deleteById(id);
    }
}

File: Python_Angular_Typescript/02 - Challenge/Resolved/back/src/service/movie_service.py
from typing import List
from src.model.movie import Movie
from src.repository.movie_repository import MovieRepository
from fastapi import HTTPException

class MovieService:
    def __init__(self):
        self.movie_repository = MovieRepository()

    def get_all_movies(self) -> List[Movie]:
        return self.movie_repository.find_all()

    def get_movie_by_id(self, movie_id: int) -> Movie:
        movie = self.movie_repository.find_by_id(movie_id)
        if not movie:
            raise HTTPException(status_code=404, detail="Movie not found")
        return movie

    def save_movie(self, movie: Movie) -> Movie:
        if self.movie_repository.exists_by_title_and_director(movie.title, movie.director):
            raise HTTPException(status_code=409, detail="Movie already exists")
        return self.movie_repository.save(movie)

    def update_movie(self, movie_id: int, movie: Movie) -> Movie:
        existing_movie = self.movie_repository.find_by_id(movie_id)
        if not existing_movie:
            raise HTTPException(status_code=404, detail="Movie not found")
        if (existing_movie.title != movie.title or existing_movie.director != movie.director) and \
                self.movie_repository.exists_by_title_and_director(movie.title, movie.director):
            raise HTTPException(status_code=409, detail="Movie already exists")
        existing_movie.title = movie.title
        existing_movie.director = movie.director
        existing_movie.release_year = movie.release_year
        return self.movie_repository.save(existing_movie)

    def delete_movie(self, movie_id: int) -> None:
        if not self.movie_repository.find_by_id(movie_id):
            raise HTTPException(status_code=404, detail="Movie not found")
        self.movie_repository.delete_by_id(movie_id)